File: 数据结构/数据结构/03-Selection-Sort/05-Performance-of-Selection-Sort/src/SelectionSort.java
public class SelectionSort {

    private SelectionSort(){}

    public static <E extends Comparable> void sort(E[] arr){

        for(int i = 0; i < arr.length; i ++){

            // 选择 arr[i...n) 中的最小值
            int minIndex = i;
            for(int j = i; j < arr.length; j ++){
                if(arr[j].compareTo(arr[minIndex]) < 0)
                    minIndex = j;
            }

            swap(arr, i, minIndex);
        }
    }

    private static <E> void swap(E[] arr, int i, int j){

        E t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }



    public static void main(String[] args){

        int[] dataSize = {10000, 100000};
        for(int n: dataSize){
            Integer[] arr = ArrayGenerator.generateRandomArray(n, n);
            SortingHelper.sortTest("SelectionSort", arr);
        }
    }
}

File: 9_排序和搜索/template/SelectionSort.ts
// 每一趟从待排序的数据元素中选择最小（或最大）的一个元素作为首元素，直到所有元素排完为止
const selectSort = (arr: number[]): void => {
  if (arr.length <= 1) return

  for (let i = 0; i < arr.length - 1; i++) {
    let minIndex = i
    for (let j = i; j < arr.length; j++) {
      if (arr[j] < arr[minIndex]) {
        minIndex = j
      }
    }

    const tmp = arr[minIndex]
    arr[minIndex] = arr[i]
    arr[i] = tmp
  }
}

export {}

if (require.main === module) {
  const arr = [4, 1, 2, 5, 3, 6, 7]
  selectSort(arr)
  console.log(arr)
}
